csrc/welford.cu

```diff
@@ -0,0 +1,429 @@
+#include <iostream>
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include <vector>
+
+
+__device__ __forceinline__ int lastpow2(int n)
+{
+  int out = 1 << (31 - __clz(n));
+  if(n == out) 
+    out >>= 1;
+  return out;
+}
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void welford_kernel(
+      const scalar_t* __restrict__ input,
+      scalar_t* __restrict__ out_mean,
+      scalar_t* __restrict__ out_var,
+      scalar_t* __restrict__ out_var_biased,
+      const int bs,
+      const int fs,
+      const int ss) {
+  extern __shared__ int s_mem[];
+  int block_size = blockDim.x * blockDim.y;
+
+  float *mean_l = (float*) s_mem;
+  float *m2n_l = (float*) &(s_mem[block_size]);
+  int *num_item_l = (int*) &(s_mem[block_size*2]);
+
+  int count = 0;
+  float x_mean = 0;
+  float m_2_n = 0;
+  int input_base = blockIdx.x*ss + threadIdx.y*ss*fs;
+  int thread_id = threadIdx.y*blockDim.x + threadIdx.x;
+
+  // sequential welford
+  for (int offset = threadIdx.x; offset < ss ; offset+= blockDim.x) {
+    count++;
+    auto x_n = static_cast<accscalar_t>(input[offset+input_base]);
+    auto x_mean_new = x_mean + (x_n - x_mean) / count;
+    m_2_n = m_2_n + (x_n - x_mean_new) * (x_n - x_mean);
+    x_mean = x_mean_new;
+  }
+
+  // allow idle thread to write to shared memory
+  mean_l[thread_id] = x_mean;
+  m2n_l[thread_id] = m_2_n;
+  num_item_l[thread_id] = count;
+  __syncthreads();
+  
+  // parallel reduce with interleaved threads
+  // TODO(jie): unroll this?
+  // TODO(jie): maybe I should pad the blockDim.y to power of 2?
+  for (int offset = lastpow2(block_size); offset > 0; offset>>=1) {
+    // excluding idle threads, because /0!
+    if (thread_id < offset && thread_id + offset < block_size && threadIdx.x < ss) {
+      auto count = num_item_l[thread_id];
+      auto val = mean_l[thread_id];
+      auto count2 = num_item_l[thread_id+offset];
+      auto val2 = mean_l[thread_id+offset];
+
+      mean_l[thread_id] = (val * count + val2 * count2) / (count + count2);
+      val = val - val2;
+      m2n_l[thread_id] += m2n_l[thread_id + offset] + val*val*count*count2/(count+count2);
+      num_item_l[thread_id] = count + count2;
+    }
+    __syncthreads();
+  }
+
+  if (thread_id == 0) {
+    out_mean[blockIdx.x] = static_cast<scalar_t>(mean_l[0]);
```

Comment:
intermediates for half data are written out as fp32 in cudnn. It may well be that it's not needed, but would be good to keep your options open may be with another template parameter.

Reply:
I agree it would be much better to leave it as an option.
I originally leave it as fp16 since bn running_mean/var uses fp16. But in distributed syncBN, intermediate mean/var should be all scattered to compute global mean/var, so it might be safer to output fp32 when running with distributed.

Reply:
finished most other parts. Have some questions with the intermediate precision. Will take it offline tomorrow :)